enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,603 @@
+success/failure of that: re-balances the egress assignment to other nodes (thus
+avoiding packets being sent to a egress node which is currently un-reachable).
+Going forward, both network plugins will come to support traffic load-balancing
+and a more sophisticated route health detection. That means that the underlying
+networking solution (OVS in the case of both OVN-Kubernetes and openshift-sdn)
+will manage the route health check. It is worth noting that the "amount of
+re-balancing" done by the network plugin is expected to be reduced, since the
+network fabric will be able to load balance traffic between all egress nodes
+provided and always use a healthy node (thus not requiring the network plugin to
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+#### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+##### annotated approach
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`cloud.openshift.io/node-ifaddr: {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}`
+
+This will in turn be read by the network plugin and used during the egress IP
+assignment, superseding any host level subnet information already parsed by the
+network plugin.
+
+##### CRD
+
+The cloud component will define a dedicated CRD that the network plugin will use
+to issue cloud egress IP requests. The network plugin will in-turn finalize its
+setup and set its final egress IP status pending the CR's status indicating a
+successful/unsuccessful cloud assignment. For example: if the cloud assignment
+should be unsuccessful; it will be up to the network plugin to notify the user
+of this, by setting its own status field for the egress IP assignment
+accordingly.
+
+The CRD will have `scope: Cluster` (since egress IP is cluster-scoped and node
+private IP assignment have no relation to namespaces) and have the following
+format:
+
+```yaml
+apiVersion: cloud.openshift.io/v1
+kind: CloudPrivateIPConfig
+metadata:
+  name: cloud-egressip
+spec:
+  items:
+  - node: nodeX
+    ip: 192.168.126.11
+  - node: nodeY
+    ip: 192.168.126.102
```

Comment:
I agree with the "single CRD with a list of nodes" form rather than the "CRD per node", because this is logically a single thing: the network plugin computes a complete allocation of egress IPs to nodes for the entire cluster, and then requests that the cloud IP provider implement that allocation. You can't process it on a node-by-node basis; if we had separate objects for nodeX and nodeY and then you swapped the two IPs around (so nodeX had 192.168.126.102 and nodeY had 191.168.126.11) then when the cloud IP provider got the update to the first node, it would be in an inconsistent state (with the same IP attached to two nodes).

It should be `name: cluster` though, for consistency with other cluster-wide singletons.

Reply:
That seems like good context for why to be single CRD with a list, perhaps some of this can be added to the doc in case others raise the same question?

Reply:
OK, I didn't have the same idea as @danwinship. I thought of having multiple CRs (one for each network plugin related egress IP assignment). But indeed: it would reduce the amount of objects to manage if we only had one which is cluster-wide as a list, and reduce the complexity w.r.t the managing multiple requests concurrently. 

> perhaps some of this can be added to the doc in case others raise the same question?

Absolutely, I will update the enhancement proposal with this.